### PR TITLE
Add an integration test for presence of expected metrics

### PR DESCRIPTION
**Pull request by swiatekm-sumo**
##### Description

Add tests for the presence of metrics we expect to send to the receiver. Uses receiver-mock's `/metrics-list` endpoint. 

Adding it to the existing Features for now, as I don't want to duplicate setup in a new Feature, and it's not yet clear how we want to deal with this structurally.

### CODE REVIEW

File: tests/integration/helm_otc_metadata_installation_test.go [hunk 1]
```diff
@@ -139,6 +141,30 @@ func Test_Helm_OT_Metadata(t *testing.T) {
 				require.EqualValues(t, 0, daemonsets[0].Status.NumberUnavailable)
 				return ctx
 			}).
+		Assess("metrics are present",
+			func(ctx context.Context, t *testing.T, envConf *envconf.Config) context.Context {
+				kubectlOpts := *ctxopts.KubectlOptions(ctx)
+				kubectlOpts.Namespace = "receiver-mock"
+				k8s.WaitUntilServiceAvailable(t, &kubectlOpts, "receiver-mock", int(waitDuration), tickDuration)
+				baseUrl := fmt.Sprintf("http://%s:%d", internal.NodePortHostname, internal.ReceiverMockNodePort)
```
Comment (pmalek-sumo) [inline tests/integration/helm_otc_metadata_installation_test.go:149]: Maybe we could think about using tunnel: https://github.com/gruntwork-io/terratest/blob/master/modules/k8s/tunnel_test.go

Comment (swiatekm-sumo) [reply tests/integration/helm_otc_metadata_installation_test.go:149]: Huh I thought that was a lot more complex for some reason. Let me try this way.

File: tests/integration/internal/constants.go [hunk 2]
```diff
@@ -13,12 +13,80 @@ const (
...
+		"kube_pod_status_phase",
+	}
+	KubeletMetrics = []string{
+		"kubelet_running_containers",
+		"kubelet_running_pods",
+	}
+	CAdvisorMetrics = []string{
+		"container_cpu_usage_seconds_total",
+		// "container_fs_usage_bytes",
+		// "container_fs_limit_bytes",
```
Comment (pmalek-sumo) [inline tests/integration/internal/constants.go:63]: I believe this one was is related to https://github.com/google/cadvisor/pull/2872

File: tests/integration/internal/receivermock/receiver_mock.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+	"testing"
+
+	http_helper "github.com/gruntwork-io/terratest/modules/http-helper"
+)
+
+// A HTTP client for the receiver-mock API
+type ReceiverMockClient struct {
+	baseUrl   url.URL
+	tlsConfig tls.Config
+	t         *testing.T
```
Comment (pmalek-sumo) [inline tests/integration/internal/receivermock/receiver_mock.go:18]: I'd prefer not to store this. Instead we could pass it as first param to exposed public functions (like in other "frameworks")

File: tests/integration/internal/receivermock/receiver_mock.go [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+	metricNameToCount := make(map[string]int)
+	lines := strings.Split(rawMetricsValues, "\n")
+	for _, line := range lines {
+		if len(line) == 0 {
+			continue
+		}
+		// the last colon of the line is the split point
+		splitIndex := strings.LastIndex(line, ":")
+		if splitIndex == -1 || splitIndex == 0 {
+			return nil, fmt.Errorf("failed to parse metrics list line: '%s'", line)
```
Comment (pmalek-sumo) [inline tests/integration/internal/receivermock/receiver_mock.go:59]:
```suggestion
			return nil, fmt.Errorf("failed to parse metrics list line: %q", line)
```

equivalent

File: tests/integration/internal/receivermock/receiver_mock.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+	baseUrl   url.URL
+	tlsConfig tls.Config
+	t         *testing.T
+}
+
+func NewReceiverMockClient(t *testing.T, baseUrl url.URL) *ReceiverMockClient {
+	return &ReceiverMockClient{baseUrl: baseUrl, t: t, tlsConfig: tls.Config{}}
+}
+
+func (client *ReceiverMockClient) GetMetricCounts() (map[string]int, error) {
```
Comment (pmalek-sumo) [inline tests/integration/internal/receivermock/receiver_mock.go:25]: Might want to define a type for this return type

```go
type MetricCounts map[string]int
```

File: tests/integration/internal/stepfuncs/assess_funcs.go [hunk 6]
```diff
@@ -2,25 +2,78 @@ package stepfuncs
...
+		require.Eventually(t, func() bool {
+			metricCounts, err := client.GetMetricCounts()
+			if err != nil {
+				t.Log(err)
+				return false
+			}
+			for _, expectedMetricName := range expectedMetrics {
+				_, ok := metricCounts[expectedMetricName]
+				if !ok {
+					t.Logf("Couldn't find metric '%s' in %v", expectedMetricName, metricCounts)
```
Comment (pmalek-sumo) [inline tests/integration/internal/stepfuncs/assess_funcs.go:71]:
```suggestion
					t.Logf("Couldn't find metric %q in %v", expectedMetricName, metricCounts)
```

File: tests/integration/internal/stepfuncs/assess_funcs.go [hunk 7]
```diff
@@ -2,25 +2,78 @@ package stepfuncs
...
+					return false
+				}
+			}
+			return true
+		}, waitDuration, tickDuration*5)
```
Comment (pmalek-sumo) [inline tests/integration/internal/stepfuncs/assess_funcs.go:76]: `tickDuration*5` seems a bit magic to me.

File: tests/integration/helm_otc_metadata_installation_test.go [hunk 8]
```diff
@@ -146,6 +147,16 @@ func Test_Helm_OT_Metadata(t *testing.T) {
 				require.EqualValues(t, 0, daemonsets[0].Status.NumberUnavailable)
 				return ctx
 			}).
+		Assess("metrics are present",
```
Comment (pmalek-sumo) [inline tests/integration/helm_otc_metadata_installation_test.go:150]: Let's add a TODO to extract this out to a separate feature when we figure out sharing a cluster between features.

**Review by pmalek-sumo — CHANGES_REQUESTED** [review]

File: tests/integration/internal/constants.go (hunk 2 above)
Comment (swiatekm-sumo) [reply tests/integration/internal/constants.go:63]: So it'll start working after kind gets to containerd 1.5.7: https://github.com/containerd/containerd/issues/5882? I'll add a comment.

File: tests/integration/internal/stepfuncs/assess_funcs.go (hunk 7 above)
Comment (swiatekm-sumo) [reply tests/integration/internal/stepfuncs/assess_funcs.go:76]: Yeah, that's a debugging holdover that shouldn't be here. Good catch!

File: tests/integration/helm_otc_metadata_installation_test.go [hunk 9]
```diff
@@ -145,7 +146,17 @@ func Test_Helm_OT_Metadata(t *testing.T) {
 
 				require.EqualValues(t, 0, daemonsets[0].Status.NumberUnavailable)
 				return ctx
-			}).
+						}).
```
Comment (pmalek-sumo) [inline tests/integration/helm_otc_metadata_installation_test.go:149]: This indentation looks weird

Comment (swiatekm-sumo) [reply tests/integration/helm_otc_metadata_installation_test.go]: Weird, I swear go fmt did this...

File: tests/integration/internal/receivermock/receiver_mock.go [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+	"fmt"
+	"net/url"
+	"strconv"
+	"strings"
+	"testing"
+
+	http_helper "github.com/gruntwork-io/terratest/modules/http-helper"
+)
+
+type MetricCounts map[string]int
```
Comment (pmalek-sumo) [inline tests/integration/internal/receivermock/receiver_mock.go:14]: Add a comment on this type what does it represent: I assume it's a mapping of metric names to metric counts but who knows :)

**Review by pmalek-sumo — APPROVED** [review]
Apart from one nit 👍

**Pull request merged**